pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/pdf/DownloadPDF.java

```diff
@@ -84,39 +86,60 @@ public DownloadPDF(@NonNull String requestBody) {
+                    maybePdf = getPDFBundle(ddpInstance, requestBody, user);
                 } else {
-                    pdfBytes = generateSinglePDF(requestBody, configName, user, ddpInstance);
+                    maybePdf = generateSinglePDF(requestBody, configName, user, ddpInstance);
+                }
+                if (maybePdf.isPresent()) {
+                    byte[] pdfByte = maybePdf.get();
+                    if (pdfByte.length > 0) {
+                        savePDFinBucket(ddpInstance.getName(), ddpParticipantId, new ByteArrayInputStream(pdfByte),
+                                fileName, user.getId());
+                    } else {
+                        throw new RuntimeException("PDF size was zero!" + fileName);
+                    }
                 }
-                pdfBytes.ifPresent(bytes -> {
-                    savePDFinBucket(ddpInstance.getName(), ddpParticipantId, new ByteArrayInputStream(bytes), fileName, user.getId());
-                });
-                pdfBytes.orElseThrow();
+                maybePdf.orElseThrow();
+
             } else {
                 throw new RuntimeException("DDPInstance of participant " + ddpParticipantId + " not found");
             }
         }
-        return pdfBytes;
+        return maybePdf;
     }
 
     private Optional<byte[]> generateSinglePDF(@NonNull String requestBody, String configName, UserDto user, DDPInstance ddpInstance) {
         byte[] pdfByte = null;
-        if (COVER.equals(configName)) {
-            pdfByte = new MRCoverPDF(this).getMRCoverPDF(requestBody, ddpInstance, user);
-        } else if (IRB.equals(configName)) {
-            String groupId = DDPInstance.getDDPGroupId(ddpInstance.getName());
-            pdfByte = PDFProcessor.getTemplateFromGoogleBucket(groupId + "_IRB_Letter.pdf");
-        } else if (TISSUE.equals(configName) || REQUEST.equals(configName)) {
-            TissueCoverPDF tissueCoverPDF = new TissueCoverPDF(this);
-            pdfByte = tissueCoverPDF.getTissueCoverPDF(ddpInstance, user);
-        } else if (configName != null) {
-            pdfByte = requestPDF(ddpInstance, ddpParticipantId, configName);
+        int turn = 0;
+        while (turn < 5) {
+            if (COVER.equals(configName)) {
+                pdfByte = new MRCoverPDF(this).getMRCoverPDF(requestBody, ddpInstance, user);
+            } else if (IRB.equals(configName)) {
+                String groupId = DDPInstance.getDDPGroupId(ddpInstance.getName());
+                pdfByte = PDFProcessor.getTemplateFromGoogleBucket(groupId + "_IRB_Letter.pdf");
+            } else if (TISSUE.equals(configName) || REQUEST.equals(configName)) {
+                TissueCoverPDF tissueCoverPDF = new TissueCoverPDF(this);
+                pdfByte = tissueCoverPDF.getTissueCoverPDF(ddpInstance, user);
+            } else if (configName != null) {
+                pdfByte = requestPDF(ddpInstance, ddpParticipantId, configName);
+            }
```

Comment:
(My previous comment might have not gone through?) Make sure you have logging that tells you which case is causing the pdfByte to be an empty array.

Reply:
Your approach seems to be to retry getting pdfs at the top-level. But at the lower level looks like you have multiple ways that those PDFS are generated (i.e., grabbed from a bucket, generated locally or fetched from DSS). Make sure you know which one is causing you trouble.
If it is the DSS fetching, then wonder if you should have a retry at the level of calling DSS? Would also be interested in seeing the actual error messages. Could very well be that there is a problem on the DSS side.
I mean: in distributed systems you always need to assume that stuff will fail, and retries are one strategy to deal with it. But you would hope that the retries are mostly needed in rare occasions!

Reply:
@MocanaAtBroad our suspicion is that we are getting pdfs from DSS but that they are size 0. this is why we are now adding a check of length > 0 and the retry if it was so. 
@pegahtah but I agree that I would like to know when this was the case and for which pdf. Can you please add in the else also a log.error: "[PDF] was size 0", Thank you. Error for now and when we know if identified the problem further we can change it from error to warn later (if it is to chatty…)